### ISSUE TITLE
Refactoring of variables handling in dune files

### ISSUE THREAD

**Issue by jeremiedimino** [ocaml/dune#1888]
Currently, variables are resolved and interpreted very late in the pipeline. It is done manually in various parts of the code while interpreting the parsed stanzas in order to produce rules. Given that we only have static scoping in dune, I propose to make the resolution of variable names more uniform, and in particular do it right during parsing. That should make it easier to implement other features such as local let-bindings or do static analysis of dune files.

One example I have in mind is the following, which I have seen come up several time:

```scheme
(let x
 (cond
  (%{ocaml-version} <= 4.06 -> 406)
  (%{ocaml-version} = 4.07 -> 407)
  (%{ocaml-version} >= 4.08 -> 408)))

(rule (copy file.%{x}.ml file.ml))
```

It would be tedious to properly add support for such variables without this PR.

Note: the syntax in this example is just an example and is not relevant for this PR.

## Overview

Parsed atoms that contain variables are always represented as `String_with_vars.t` values. Such values represent variables by their name. For instance `%{ocaml-version}` is represented as `Var "ocaml-version"`. I propose to represent it by `Var Ocaml_version` instead, which means that the variable will be resolved right during parsing. During evaluation, we will just have to associate values to already resolved variables.

## Implementation

During parsing, we maintain a set of variables in scope, which is basically a mapping from names to the constructor representing the resolved variable. This mapping is stored in the `Univ_map.t` context of the parser and helpers are provided to add new variables to the context:

```
val bind
  :  string
  -> Pform_repr.t
  -> ('a, 'b) Dune_lang.Decoder.parser
  -> ('a, 'b) Dune_lang.Decoder.parser
```

The decoder for `String_with_vars.t` will then immediately resolve all the variables at parsing time using this mapping.

**Comment by rgrinberg** [comment]
> Parsed atoms that contain variables are always represented as String_with_vars.t values. Such values represent variables by their name. For instance %{ocaml-version} is represented as Var "ocaml-version". I propose to represent it by Var Ocaml_version instead, which means that the variable will be resolved right during parsing. During evaluation, we will just have to associate values to already resolved variables.

How does this help exactly however. If we are to introduce user defined variables, we'll still need some sort of string variables for user defined variables.

Handling variable resolution in the parser itself seems good however.

**Comment by diml** [comment]
I suppose user variables would be represented by something like `User_var of User_var.Id.t`

**Comment by rgrinberg** [comment]
To be more concrete, let's see how this is going to work for rule defined bindings for example:

```
(rule
 (deps (:foo bar))
 (targets baz)
 (action (run ./exe %{foo} -o %{targets})))
```

Here, we need to make sure to parse `deps` ahead of `action`. When we parse `deps` we must make sure to add mappings for every binding introduced using `bind`. Consequently, when we parse `action` we use the map to resolve some of these bindings.

I assume that `Pform_repr.t` is going to be something that will make sure that subsequent variable references `%{foo}` refer to the same definition?

**Comment by diml** [comment]
Yes, that's the idea.

**Comment by rgrinberg** [comment]
Jeremie implemented this in 3.0
